File: ya/deepseek.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable
import json
import time
import urllib.error
import urllib.request

from .config import ModelConfig


API_URL = "https://api.deepseek.com/chat/completions"
MAX_TOOL_CALL_ROUNDS = 6
MAX_REQUEST_ATTEMPTS = 3


class DeepSeekError(RuntimeError):
    pass


@dataclass
class ModelReply:
    content: str
    reasoning_content: str | None
    tool_calls: list[dict]
    usage: dict
    assistant_message: dict


class DeepSeekClient:
    def __init__(
        self,
        api_key: str,
        opener: Callable = urllib.request.urlopen,
        sleep: Callable[[float], None] = time.sleep,
    ):
        self.api_key = api_key
        self.opener = opener
        self.sleep = sleep

    def _payload(
        self,
        messages: list[dict],
        config: ModelConfig,
        max_tokens: int,
        tools: list[dict] | None,
        stream: bool,
    ) -> dict:
        payload = {
            "model": config.model,
            "messages": messages,
            "thinking": {"type": "enabled" if config.thinking_enabled else "disabled"},
            "stream": stream,
            "max_tokens": max_tokens,
        }
        if config.thinking_enabled:
            payload["reasoning_effort"] = config.reasoning_effort
        if tools:
            payload["tools"] = tools
        if stream:
            payload["stream_options"] = {"include_usage": True}
        return payload

    def _request(self, payload: dict, timeout: int):
        request = urllib.request.Request(
            API_URL,
            data=json.dumps(payload).encode("utf-8"),
            headers={"Authorization": f"Bearer {self.api_key}", "Content-Type": "application/json"},
            method="POST",
        )
        for attempt in range(MAX_REQUEST_ATTEMPTS):
            try:
                return self.opener(request, timeout=timeout)
            except urllib.error.HTTPError as error:
                if error.code < 500 or attempt == MAX_REQUEST_ATTEMPTS - 1:
                    detail = error.read().decode("utf-8", errors="replace")
                    raise DeepSeekError(f"DeepSeek API returned HTTP {error.code}: {detail}") from error
            except urllib.error.URLError as error:
                if attempt == MAX_REQUEST_ATTEMPTS - 1:
                    raise DeepSeekError(f"DeepSeek API request failed: {error.reason}") from error
            self.sleep(0.5 * (attempt + 1))
        raise AssertionError("unreachable")

    def complete(self, messages: list[dict], config: ModelConfig, max_tokens: int, tools: list[dict] | None = None) -> ModelReply:
        config.validate()
        payload = self._payload(messages, config, max_tokens, tools, stream=False)
        with self._request(payload, config.toa_timeout) as response:
            body = json.loads(response.read().decode("utf-8"))

        try:
            message = body["choices"][0]["message"]
        except (KeyError, IndexError, TypeError) as error:
            raise DeepSeekError(f"Unexpected DeepSeek response: {body}") from error
        return ModelReply(
            content=message.get("content") or "",
            reasoning_content=message.get("reasoning_content"),
            tool_calls=message.get("tool_calls") or [],
            usage=body.get("usage") or {},
            assistant_message=message,
        )

    def complete_stream(
        self,
        messages: list[dict],
        config: ModelConfig,
        max_tokens: int,
        on_content: Callable[[str], None],
    ) -> ModelReply:
        """Stream a tool-free final answer and return its accumulated reply."""
        config.validate()
        payload = self._payload(messages, config, max_tokens, tools=None, stream=True)
        content: list[str] = []
        reasoning: list[str] = []
        usage: dict = {}
        # Retry only before any text is emitted, so a reconnect cannot duplicate output.
        for attempt in range(MAX_REQUEST_ATTEMPTS):
            try:
                with self._request(payload, config.toa_timeout) as response:
                    for raw_line in response:
                        line = raw_line.decode("utf-8", errors="replace").strip()
                        if not line.startswith("data:"):
                            continue
                        data = line[5:].strip()
                        if data == "[DONE]":
                            return ModelReply(
                                content="".join(content),
                                reasoning_content="".join(reasoning) or None,
                                tool_calls=[],
                                usage=usage,
                                assistant_message={"role": "assistant", "content": "".join(content)},
                            )
                        chunk = json.loads(data)
                        usage.update(chunk.get("usage") or {})
                        delta = (chunk.get("choices") or [{}])[0].get("delta") or {}
                        hidden_reasoning = delta.get("reasoning_content")
                        if hidden_reasoning:
                            reasoning.append(hidden_reasoning)
                        text = delta.get("content")
                        if text:
                            content.append(text)
                            on_content(text)
                return ModelReply(
                    content="".join(content),
                    reasoning_content="".join(reasoning) or None,
                    tool_calls=[],
                    usage=usage,
                    assistant_message={"role": "assistant", "content": "".join(content)},
                )
            except (urllib.error.URLError, urllib.error.HTTPError) as error:
                if content or attempt == MAX_REQUEST_ATTEMPTS - 1:
                    if isinstance(error, urllib.error.HTTPError):
                        detail = error.read().decode("utf-8", errors="replace")
                        raise DeepSeekError(f"DeepSeek API returned HTTP {error.code}: {detail}") from error
                    raise DeepSeekError(f"DeepSeek API request failed: {error.reason}") from error
                self.sleep(0.5 * (attempt + 1))
        raise AssertionError("unreachable")

    def run_with_tools(
        self,
        messages: list[dict],
        config: ModelConfig,
        max_tokens: int,
        tools: list[dict] | None = None,
        tool_handlers: dict[str, Callable[[dict], str]] | None = None,
    ) -> ModelReply:
        """Preserve reasoning only inside a single tool-call turn."""
        transient_messages = list(messages)
        handlers = tool_handlers or {}
        for round_number in range(MAX_TOOL_CALL_ROUNDS + 1):
            reply = self.complete(transient_messages, config, max_tokens, tools)
            if not reply.tool_calls:
                return reply
            if round_number == MAX_TOOL_CALL_ROUNDS:
                break
            transient_messages.append(reply.assistant_message)
            for call in reply.tool_calls:
                function = call.get("function", {})
                name = function.get("name")
                if name not in handlers:
                    result = json.dumps({"error": f"Tool {name!r} is not available."})
                else:
                    try:
                        arguments = json.loads(function.get("arguments") or "{}")
                        result = handlers[name](arguments)
                    except Exception as error:  # Tool errors are returned to the model, not persisted.
                        result = json.dumps({"error": str(error)})
                transient_messages.append({"role": "tool", "tool_call_id": call["id"], "content": result})
        # Some models keep searching after the useful evidence is already available.
        # Give them one tool-free turn to synthesize an answer instead of failing the CLI.
        transient_messages.append(
            {
                "role": "user",
                "content": (
                    "The tool-call budget is exhausted. Return the best final answer to the "
                    "original request now without using any tools. Be concise and state "
                    "uncertainty when needed."
                ),
            }
        )
        reply = self.complete(transient_messages, config, max_tokens, tools=None)
        if not reply.tool_calls:
            return reply
        raise DeepSeekError(
            f"Tool-call limit ({MAX_TOOL_CALL_ROUNDS} rounds) reached before a final answer."
        )

File: src/deepseek.ts
import {
  SseReader,
  buildChatPayload,
  parseModelReply,
  parseStreamChunk,
} from "ya-core";
import type {
  ChatMessage,
  TokenUsage,
  ToolArguments,
  ToolCall,
  ToolDefinition,
  ToolHandler,
} from "./types";
import { ModelConfig } from "./config";

export const API_URL = "https://api.deepseek.com/chat/completions";
export const MAX_TOOL_CALL_ROUNDS = 6;
export const MAX_REQUEST_ATTEMPTS = 3;

export type FetchLike = (input: string | URL | Request, init?: RequestInit) => Promise<Response>;
export type Sleep = (milliseconds: number) => Promise<void>;

const defaultSleep: Sleep = (milliseconds) => new Promise((resolve) => setTimeout(resolve, milliseconds));

export class DeepSeekError extends Error {
  override readonly name = "DeepSeekError";
}

export interface ModelReply {
  content: string;
  reasoningContent?: string;
  toolCalls: ToolCall[];
  usage: TokenUsage;
  assistantMessage: ChatMessage;
}

interface DeepSeekPayload {
  model: string;
  messages: ChatMessage[];
  thinking: { type: "enabled" | "disabled" };
  stream: boolean;
  max_tokens: number;
  reasoning_effort?: string;
  tools?: ToolDefinition[];
  stream_options?: { include_usage: true };
}

export class DeepSeekClient {
  readonly apiKey: string;
  private readonly fetcher: FetchLike;
  private readonly sleep: Sleep;

  constructor(apiKey: string, fetcher: FetchLike = fetch, sleep: Sleep = defaultSleep) {
    this.apiKey = apiKey;
    this.fetcher = fetcher;
    this.sleep = sleep;
  }

  /**
   * Builds the chat-completions body.
   *
   * The shape, the image role rule, and the per-request image limit live in the
   * Rust core; this only marshals the arguments across the boundary.
   */
  payload(
    messages: ChatMessage[],
    config: ModelConfig,
    maxTokens: number,
    tools: ToolDefinition[] | undefined,
    stream: boolean,
  ): DeepSeekPayload {
    return JSON.parse(buildChatPayload(
      JSON.stringify(messages),
      config.model,
      config.thinkingEnabled,
      config.reasoningEffort,
      maxTokens,
      tools === undefined ? undefined : JSON.stringify(tools),
      stream,
    )) as DeepSeekPayload;
  }

  private async requestOnce(payload: DeepSeekPayload, timeoutSeconds: number): Promise<Response> {
    return this.fetcher(API_URL, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(timeoutSeconds * 1_000),
    });
  }

  private async httpError(response: Response): Promise<DeepSeekError> {
    const detail = await response.text();
    return new DeepSeekError(`DeepSeek API returned HTTP ${response.status}: ${detail}`);
  }

  private networkError(error: unknown): DeepSeekError {
    return new DeepSeekError(`DeepSeek API request failed: ${error instanceof Error ? error.message : String(error)}`);
  }

  private async request(payload: DeepSeekPayload, timeoutSeconds: number): Promise<Response> {
    let lastError: unknown;
    for (let attempt = 0; attempt < MAX_REQUEST_ATTEMPTS; attempt += 1) {
      try {
        const response = await this.requestOnce(payload, timeoutSeconds);
        if (response.ok) return response;
        const error = await this.httpError(response);
        if (response.status < 500 || attempt === MAX_REQUEST_ATTEMPTS - 1) throw error;
        lastError = error;
      } catch (error) {
        if (error instanceof DeepSeekError && (error.message.includes("HTTP 4") || attempt === MAX_REQUEST_ATTEMPTS - 1)) throw error;
        lastError = error;
        if (attempt === MAX_REQUEST_ATTEMPTS - 1) throw this.networkError(error);
      }
      await this.sleep(500 * (attempt + 1));
    }
    throw this.networkError(lastError);
  }

  async complete(
    messages: ChatMessage[],
    config: ModelConfig,
    maxTokens: number,
    tools?: ToolDefinition[],
  ): Promise<ModelReply> {
    config.validate();
    const response = await this.request(this.payload(messages, config, maxTokens, tools, false), config.toaTimeout);
    // Parsing the body here keeps a non-JSON response failing exactly where it
    // did before; the Rust core receives the same re-serialised JSON that the
    // TypeScript original embedded in its error message.
    const body = await response.json() as unknown;
    return JSON.parse(parseModelReply(JSON.stringify(body))) as ModelReply;
  }

  async completeStream(
    messages: ChatMessage[],
    config: ModelConfig,
    maxTokens: number,
    onContent: (content: string) => void,
  ): Promise<ModelReply> {
    config.validate();
    const payload = this.payload(messages, config, maxTokens, undefined, true);
    let lastError: unknown;
    for (let attempt = 0; attempt < MAX_REQUEST_ATTEMPTS; attempt += 1) {
      const content: string[] = [];
      const reasoning: string[] = [];
      const usage: TokenUsage = {};
      try {
        const response = await this.requestOnce(payload, config.toaTimeout);
        if (!response.ok) {
          const error = await this.httpError(response);
          if (response.status < 500 || attempt === MAX_REQUEST_ATTEMPTS - 1) throw error;
          lastError = error;
          await this.sleep(500 * (attempt + 1));
          continue;
        }
        const done = await readServerSentEvents(response, (data) => {
          const chunk = parseStreamChunk(data);
          if (chunk.done) return true;
          if (chunk.usageJson) Object.assign(usage, JSON.parse(chunk.usageJson) as TokenUsage);
          if (chunk.reasoningContent != null) reasoning.push(chunk.reasoningContent);
          if (chunk.content != null) {
            content.push(chunk.content);
            onContent(chunk.content);
          }
          return false;
        });
        void done;
        const joinedContent = content.join("");
        const joinedReasoning = reasoning.join("");
        return {
          content: joinedContent,
          ...(joinedReasoning ? { reasoningContent: joinedReasoning } : {}),
          toolCalls: [],
          usage,
          assistantMessage: { role: "assistant", content: joinedContent },
        };
      } catch (error) {
        lastError = error;
        if (content.length > 0 || attempt === MAX_REQUEST_ATTEMPTS - 1 || error instanceof DeepSeekError && error.message.includes("HTTP 4")) {
          throw error instanceof DeepSeekError ? error : this.networkError(error);
        }
        await this.sleep(500 * (attempt + 1));
      }
    }
    throw this.networkError(lastError);
  }

  async runWithTools(
    messages: ChatMessage[],
    config: ModelConfig,
    maxTokens: number,
    tools?: ToolDefinition[],
    toolHandlers: Record<string, ToolHandler> = {},
  ): Promise<ModelReply> {
    const transientMessages = [...messages];
    for (let round = 0; round <= MAX_TOOL_CALL_ROUNDS; round += 1) {
      const reply = await this.complete(transientMessages, config, maxTokens, tools);
      if (reply.toolCalls.length === 0) return reply;
      if (round === MAX_TOOL_CALL_ROUNDS) break;
      transientMessages.push(reply.assistantMessage);
      for (const call of reply.toolCalls) {
        let result: string;
        const handler = toolHandlers[call.function.name];
        if (!handler) {
          result = JSON.stringify({ error: `Tool '${call.function.name}' is not available.` });
        } else {
          try {
            const parsed = JSON.parse(call.function.arguments || "{}") as unknown;
            if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
              throw new Error("Tool arguments must be a JSON object.");
            }
            result = await handler(parsed as ToolArguments);
          } catch (error) {
            result = JSON.stringify({ error: error instanceof Error ? error.message : String(error) });
          }
        }
        transientMessages.push({ role: "tool", tool_call_id: call.id, content: result });
      }
    }
    transientMessages.push({
      role: "user",
      content: "The tool-call budget is exhausted. Return the best final answer to the original request now without using any tools. Be concise and state uncertainty when needed.",
    });
    const reply = await this.complete(transientMessages, config, maxTokens);
    if (reply.toolCalls.length === 0) return reply;
    throw new DeepSeekError(`Tool-call limit (${MAX_TOOL_CALL_ROUNDS} rounds) reached before a final answer.`);
  }
}

/**
 * Reads the event stream, with the framing and the incremental UTF-8 decoding
 * handled by the Rust core so a multi-byte character split across two chunks
 * stays intact.
 */
async function readServerSentEvents(response: Response, onData: (data: string) => boolean): Promise<boolean> {
  if (!response.body) throw new DeepSeekError("DeepSeek streaming response had no body.");
  const reader = response.body.getReader();
  const frames = new SseReader();
  while (true) {
    const { done, value } = await reader.read();
    if (value) {
      for (const data of frames.push(Buffer.from(value))) {
        if (onData(data)) {
          await reader.cancel();
          return true;
        }
      }
    }
    if (done) break;
  }
  const trailing = frames.finish();
  return trailing !== null && onData(trailing);
}
